### ISSUE TITLE
#582 breaks format detection

### ISSUE THREAD

**Issue by tt6746690** [samtools/htslib#584]
I linked htslib in master to samtools. `samtools view url` where `url` points to a local server implementing htsget protocol. I received the following error, 
```sh
[E::hts_hopen] Failed to open file http://127.0.0.1:8888/reads/test?format=BAM&referenceName=chr1&start=10145&end=10150
[E::hts_open_format] Failed to open file http://127.0.0.1:8888/reads/test?format=BAM&referenceName=chr1&start=10145&end=10150
samtools view: failed to open "http://127.0.0.1:8888/reads/test?format=BAM&referenceName=chr1&start=10145&end=10150" for reading: Operation not supported
```
The 1.5 release of `htslib` and `samtools` worked fine. It seems that #582 breaks `hts_detect_format` where [hpeek](https://github.com/samtools/htslib/blob/75d9f0d9a973898686a51b60277f239c7e729ea5/hts.c#L201) returns -1. I did some debugging, but to no avail, it seems changes in `multipart.c` breaks it. Adding the following code after [here](https://github.com/samtools/htslib/blob/7cdd5749e8054527912df184be87ac7663b01115/multipart.c#L262#L262)

```c
char s[32];
ssize_t len = hpeek(&fp->base, s, 18);
fprintf(stderr, "release 1.5: htsget_redirect after reset backend len=%d, s=%s\n", len, s);
// release 1.5: htsget_redirect after reset backend len=0, s=
```
```c
char s[32];
ssize_t len = hpeek(&fp->base, s, 18);
fprintf(stderr, "with htslib on develop: htsget_redirect after reset backend len=%d, s=%s\n", len, s);
// with htslib on develop: htsget_redirect after reset backend len=-1, s=}G[
```

`len`  changed from 0 to -1. Can someone tell me why this is happening.

**Comment by daviesrob** [comment]
The htsget protocol changed slightly in commit [hts-specs/5448ce9](https://github.com/samtools/hts-specs/commit/5448ce9).  #582 implemented that change in htslib.  When the servers have been updated to output the revised JSON format it should start working again.

**Comment by valeriuo** [comment]
Also, 7cdd5749e8054527912df184be87ac7663b01115 has not been pushed to master.

**Comment by tt6746690** [comment]
The server already implements the part where json response is wrapped in a `htsget` key. The json is the stringified version of this 
```json
{
    "htsget": {
        "format": "BAM",
        "sha256": "ae0ab5c12e25079556b3be64cc6150d318ba2743f6155eedcd29e5d2d7ae2033",
        "urls": [
            {
                "headers": {
                    "Range": "bytes=0-584"
                },
                "url": "http://127.0.0.1:8888/data/cbfc6439b31af855c82491c2db08602506f0734c14e5aa32ebdb02be9a303e65"
            }
        ]
    }
}
```
The response is as follows 
```sh 
< HTTP/1.0 200 OK
< Content-Type: application/json
< Content-Length: 248
< {"htsget":{"format":"BAM", ...... 
```

Also I was assuming the default branch is `master`. As it turns out it is `develop` instead;

**Comment by daviesrob** [comment]
Ah, in that case the problem is the other way around.  The update is currently only in `develop` as you've noticed.  `master` just contains the HTSlib releases. The next one (including this update) should appear some time in September if all goes well.

**Comment by tt6746690** [comment]
Thanks. Just want to point out that this problem is there.

**Comment by daviesrob** [comment]
Closing, as master has included this change since version 1.6
